libkbfs/file_data.go

```diff
@@ -1523,37 +1835,84 @@ func (fd *fileData) deepCopy(ctx context.Context, codec kbfscodec.Codec,
 // undupChildrenInCopy takes a top block that's been copied via
 // deepCopy(), and un-deduplicates all leaf children of the block.  It
 // adds all child blocks to the provided `bps`, including both the
-// ones that were deduplicated and the ones that weren't (TODO).  It
-// returns the BlockInfos for all children, including ones that
-// weren't de-duplicated.
+// ones that were deduplicated and the ones that weren't.  It returns
+// the BlockInfos for all children.
 func (fd *fileData) undupChildrenInCopy(ctx context.Context,
 	bcache BlockCache, bops BlockOps, bps *blockPutState,
 	topBlock *FileBlock) ([]BlockInfo, error) {
 	if !topBlock.IsInd {
 		return nil, nil
 	}
 
-	// TODO: support multiple levels of indirection.
-	// TODO: parallelize
-	blockInfos := make([]BlockInfo, len(topBlock.IPtrs))
-	for i, iptr := range topBlock.IPtrs {
-		if iptr.RefNonce == kbfsblock.ZeroRefNonce {
-			// No need to un-dup mid-level indirect blocks.  TODO:
-			// fetch the blocks and add them to the bps (only needed
-			// for multiple levels of indirection).
-			blockInfos = append(blockInfos, iptr.BlockInfo)
-			continue
-		}
-		childBlock, _, err := fd.getter(ctx, fd.kmd, iptr.BlockPointer,
-			fd.file, blockRead)
+	// For indirect files, get all the paths to leaf blocks.  Note
+	// that because topBlock is a deepCopy, all of the blocks are also
+	// deepCopys and thus are modifiable.
+	pfr, err := fd.getIndirectBlocksForOffsetRange(ctx, topBlock, 0, -1)
+	if err != nil {
+		return nil, err
+	}
+	if len(pfr) == 0 {
+		return nil, fmt.Errorf(
+			"Indirect file %v had no indirect blocks", fd.rootBlockPointer())
+	}
+
+	// Append the leaf block to each path, since readyHelper expects
+	// it.
+	for i, path := range pfr {
+		leafPtr := path[len(path)-1].childIPtr().BlockPointer
+		leafBlock, _, err := fd.getter(
+			ctx, fd.kmd, leafPtr, fd.file, blockWrite)
```

Comment:
@akalin-keybase: I tend to agree this might be niche enough to release journaling with this problem, although who knows, the disk-based cache stuff might be done by the time journaling is bug-free anyway.

To be clear, the dirty and ready caches I'm talking about here would only live for the lifetime of a process, and could be cleared on restart.  They'd have no overlap with the journal.  The clean disk block cache would have some content overlap with the journal if we don't combine the two somehow, yes, but that's somewhat orthogonal to the issue here I think.